### PR TITLE
3D Bounding Box Detection

### PR DESCRIPTION
**Pull request by ajy8456**
`anchor.py` : for generate anchors in each feature map pixels

`bbox_regression` 
- (Like YOLO v3) ResNet -> feature map -> FC(confidence/translation/scale/orientation) -> 3D bounding box regression
- contain orientation loss, model, train

### CODE REVIEW

File: scripts/bb_regression.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        translation = self.translation(x)
+        scale = self.scale(x)
+        orientation = self.orientation(x)
+        orientation = orientation.view(-1, self.bins, 2)
+        orientation = F.normalize(orientation, dim=2)
+
+        return confidence, translation, scale, orientation
+
+
+def train():
```
Comment (yycho0108) [inline scripts/bb_regression.py:94]: I think we should seriously consider rewriting this `train()` function to utilize our training pipeline framework. If there are issues, we should resolve it there rather than solving similar problems in multiple different copies of the code for how we train these networks.

File: src/top/run/anchor.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+"""
```
Comment (yycho0108) [inline src/top/run/anchor.py:1]: I feel like this file should be in `top/model` rather than `top/run`. See the explanation on the package structure at the top.

**Review by yycho0108 — CHANGES_REQUESTED** [review]
Please address some of the comments that I have left, and then we could consider how we could move forwards with this PR.

**Comment by yycho0108** [comment]
@ajy8456 Let me know if serious (and difficult) merge conflicts arise due to #8 .

File: scripts/bb_regression.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import torch
+import numpy as np
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.autograd import Variable
+from top.model.backbone import resnet_fpn_backbone
+from top.run.loss_util import *
+
+
+def OrientationLoss(orient_batch, orientGT_batch, confGT_batch):
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py]: This is a mistake that occurred while trying to create a class by combining it with another function.

File: scripts/bb_regression.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+"""
+
+
+import torch
+import numpy as np
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.autograd import Variable
+from top.model.backbone import resnet_fpn_backbone
+from top.run.loss_util import *
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py]: I will move `loss_util` to `top.model`.

File: scripts/bb_regression.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+
+def OrientationLoss(orient_batch, orientGT_batch, confGT_batch):
+
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py]: `orient_batch` is multibin of orientation(e.g. discretizate the orientation angle and divide it into n overlapping bins).
`orientGT_batch` is orientation of ground truth.
`confGT_batch` is which bin corresponds to the orientation of the ground truth.

Also, orient is (cos,sin) not a angle. I will change the name like `tri_orient` 

I am going to write comment.

File: scripts/bb_regression.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    orientGT_batch = orientGT_batch[torch.arange(batch_size), indexes]
+    orient_batch = orient_batch[torch.arange(batch_size), indexes]
+
+    theta_diff = torch.atan2(orientGT_batch[:,1], orientGT_batch[:,0])
+    estimated_theta_diff = torch.atan2(orient_batch[:,1], orient_batch[:,0])
+
+    return -1 * torch.cos(theta_diff - estimated_theta_diff).mean()
+
+
+class Model(nn.Module):
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py]: Yeah, we should call it `BoundingBoxRegressionModel`.

File: scripts/bb_regression.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    def __init__(self, features=None, bins=2, w = 0.4):
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py]: Thank you for suggestion.

File: scripts/bb_regression.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        super(Model, self).__init__()
+        self.bins = bins
+        self.w = w
+        self.features = features
+
+        self.confidence = nn.Sequential(
+                    nn.Linear(512 * 7 * 7, 256),
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py:29]: I didn't know how to get the features of the feature map of a 2D detection model like Yolo, so I implemented it with hardcode in first. This is the part that needs to be corrected.

File: scripts/bb_regression.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+                )
+
+        self.orientation = nn.Sequential(
+                    nn.Linear(512 * 7 * 7, 256),
+                    nn.ReLU(True),
+                    nn.Dropout(),
+                    nn.Linear(256, 256),
+                    nn.ReLU(True),
+                    nn.Dropout(),
+                    nn.Linear(256, bins*2) # to get sin and cos
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py:46]: From my understanding, in the reference paper, after detecting 2D bounding box, they crop the image. Then the cropped image is passed into the network, so there is only one object in image.

File: scripts/bb_regression.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+    def forward(self, x):
+        x = self.features(x) # 512 x 7 x 7
+        x = x.view(-1, 512 * 7 * 7)
+        confidence = self.confidence(x)
+        translation = self.translation(x)
+        scale = self.scale(x)
+        orientation = self.orientation(x)
+        orientation = orientation.view(-1, self.bins, 2)
+        orientation = F.normalize(orientation, dim=2)
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py]: `orientation` indicate valid (cos,sin) values, trying to match for input of 'orientation_loss'.
I will write the comment and change the name of variable.

File: scripts/bb_regression.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    
+    # hyper parameters
+    epochs = 100
+    batch_size = 8
+    alpha = 0.6
+    w = 0.4
+
+    print("Loading all detected objects in dataset...")
+
+    train_path = os.path.abspath(os.path.dirname(__file__)) + '___'
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py:82]: This is an incomplete part regarding data loading.

File: scripts/bb_regression.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+              'shuffle': True,
+              'num_workers': 6}
+
+    generator = data.DataLoader(dataset, **params)
+
+    backnone = resnet_fpn_backbone()
+    model = Model(features=backnone.features).cuda()
+    optim = torch.optim.Adam(model.parameters(), lr=0.0001, momentum=0.9)
+
+    conf_loss_func = nn.CrossEntropyLoss().cuda()
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py:95]: Sorry for not being used to handling configure yet.

File: scripts/bb_regression.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    translation_loss_func = nn.MSELoss().cuda()
+    scale_loss_func = nn.MSELoss().cuda()
+    orient_loss_func = OrientationLoss
+
+    total_num_batches = int(len(dataset) / batch_size)
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py:99]: I haven't considered that part. Thank you.

File: scripts/bb_regression.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+            truth_orient = local_labels['Orientation'].float().cuda()
+            truth_conf = local_labels['Confidence'].long().cuda()
+            truth_dim = local_labels['translation'].float().cuda()
+            truth_dim = local_labels['scale'].float().cuda()
+
+            local_batch=local_batch.float().cuda()
+            [orient, conf, dim] = model(local_batch)
+
+            translation_loss = translation_loss_func(dim, truth_dim)
+            scale_loss = translation_loss_func(dim, truth_dim)
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py]: Oh, I'll fix it.

File: src/top/run/anchor.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+from itertools import product as product
+
+import torch
+
+
+# |TODO(Jiyong)|: change for Objectron setting
+@dataclass
+class AnchorSetting:
+    image_size: int = 300
+    feature_maps: List[int] = (7,)
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply src/top/run/anchor.py]: Yeah, Tuple is better.

File: src/top/run/anchor.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+class Anchor(object):
+    """Compute priorbox coordinates in center-offset form for each source feature map."""
+    def __init__(self, cfg: AnchorSetting):
+        super(Anchor, self).__init__()
+        self.image_size = cfg.image_size
+        self.feature_maps = cfg.feature_maps
+        self.min_sizes = cfg.min_sizes
+        self.max_sizes = cfg.max_sizes
+
+    def forward(self):
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply src/top/run/anchor.py]: I change the name from `forward` to `generate`.

File: src/top/run/anchor.py (hunk 2 above)
Comment (ajy8456) [reply src/top/run/anchor.py]: I think I will not use this file because the method of using anchor has changed, but I will move it to that location.

File: scripts/bb_regression.py [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+            orient_loss = orient_loss_func(orient, truth_orient, truth_conf)
+
+            truth_conf = torch.max(truth_conf, dim=1)[1]
+            conf_loss = conf_loss_func(conf, truth_conf)
+
+            loss_theta = conf_loss + w * orient_loss
```
[the remark this answers is not in the document]
Comment (ajy8456) [reply scripts/bb_regression.py:120]: According to paper, The L2 loss is not a good fit for many complex multi-modal regression problem. So, they propose new orientation loss with MultiBin architecture.
They discretize the orientation angle and divide it into n bins. For each bin, the network estimates both confidence probability that the angle lies inside the bin and residual rotation correction that needs to be applied to the orientation of the center of that bin.

**Comment by ajy8456** [comment]
To do list
- create 2D object detection part. (detection, cropping, ..., How to train?)
- generate camera projection matrix
- organize the settings(model, train, ...)
- process for label(class average of dimension, confidence for orientation bin, cos&sin of local orientation)
- callback

File: scripts/bb_regression.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yycho0108) [reply scripts/bb_regression.py]: Cool, thanks for the explanation.

File: scripts/bb_regression.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yycho0108) [reply scripts/bb_regression.py]: Cool, thanks

File: scripts/bb_regression.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yycho0108) [reply scripts/bb_regression.py:29]: Ok, we'll try to think about the best way to do this. For now it's probably ok to leave as-is.

File: scripts/bb_regression.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yycho0108) [reply scripts/bb_regression.py:46]: Ah, sounds good. Thanks for the explanation.

File: scripts/bb_regression.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (yycho0108) [reply scripts/bb_regression.py:120]: Great. Can you add those comments in the code?

**Review by yycho0108 — APPROVED** [review]
Thanks for addressing the PR comments. Not all of them were resolved but I think we should prioritize on moving forwards rather than being stuck with some of the more trivial comments. @ajy8456 Your TODOs seem good, getting those completed would probably get us a step closer to getting the bounding box regression model fully functional.
